MEMORY.md

```diff
@@ -42,132 +65,279 @@ classDiagram
+list of active page owners (VM guests).
+
+A `PlatformPageTable` is created for each running VM and is used to
+manage the 2nd-stage translation tables for the VM, with `root` pointing to
+the 16kB root page table of the paging hierarchy. This is the entity that is
+assigned a unique `PageOwnerId` for the VM and it maintains a reference to
+`PageTracker` for checking which pages are allowed to be mapped or unmapped
+from the VM.
+
+`VmPages` is a per-VM structure that has a `PlatformPageTable` and
+a `VmRegionList`, which tracks the parts of the physical address space are
+eligible to be used for confidential or shared memory. `VmPages` is used to
+coordinate the assignment, reclaim, or sharing of pages with a child VM.
+It also contains the measurement of all data pages that were mapped into the
+VM before it was finalized.
+
+`Vm` combines a page table for the VM (`VmPages`) and the per-vCPU
+(`VmCpu`) state.
 
-A [`Sv48x4`]() is created for each running VM. This tracks the page table for
-the VM (with the 4, 16k-aligned pages for hgatp referenced by `root`). This is
-the entity that is assigned a unique `PageOwnerId` for the VM and it maintains
-a reference to `PageTracker` for checking which pages are assigned to it and
-for sharing pages with child VMs it starts.
+With the exception of shared pages, which can be simultaneously mapped into
+multiple VMs, each page must either be uniquely mapped into the address space
+of the owning VM or used to hold internal state for the owning VM.
 
-[`VmPages`]() is a per-VM structure that has an `Sv48x4` and the measurement of
-all data pages that were added to the page table before it was finalized.
+# VM address space
 
-[`HostVm`]() and [`GuestVm`]() combine a page table for the guest (`VmPages`) and the
-per-cpu register state of the VM to provide functionality that runs the VM.
+A guest VM's physical address space is made up of distinct, non-overlapping
+regions which determine what kinds of pages may be mapped into the region and
+how Salus responds to page faults taken in that region. These regions are
+defined during VM initialization and their size or location cannot be changed at
+runtime. Currently there are three types:
 
-## Donating pages to guests
+- `Confidential`: Memory that is confidential to the VM. Any pages mapped in
+a confidential region are considered to be owned by the VM and are guaranteed to
+be inaccessible to the VM's parent. Data pages, or pages initialized with
+host-provided contents, are measured as they are added and may only be added
+prior to VM finalization. Zero-filled confidential pages may be added at any
+point. Page faults in a confidential region trigger an exit to the VM's parent
+with the faulting address.
 
-# Memory assignment/sharing flows for TVMs and the hypervisor
+* `Shared`: Memory that is shared with the VM's parent. Pages in a shared
+memory region are considered to be owned by the parent VM, and the parent VM
```

Comment:
s/"parent VM"/parent/g?

The parent may not necessarily be a VM.

Reply:
Hm? Every entity in the Salus model is a VM. (Well, except for the hypervisor itself, but it does not share memory and memory cannot be returned to it).